File: iPhone/BhaktiAngan/BhaktiAngan/Services/VoiceJapaCounter.swift
import AVFoundation
import Accelerate
import Combine

/// Hands-free japa counting (Voice Japa v1, approach 1 in Docs/VOICE_JAPA.md).
///
/// Taps the microphone with `AVAudioEngine`, measures short-time loudness (RMS),
/// and counts each *utterance burst* (one mantra repetition = one voiced burst
/// separated by a breath). An adaptive noise floor + hysteresis + a refractory
/// window keep it from double-counting syllables or triggering on room noise.
/// A short calibration ("chant 3 times") learns the devotee's level and rhythm.
///
/// Audio is processed live on the device and is **never recorded or transmitted**,
/// so the App Privacy "Data Not Collected" label holds. Foreground only (the
/// devotee is actively chanting), so no background-audio entitlement is needed.
///
/// NOTE: microphone counting cannot be validated in the iOS Simulator (no mic).
/// The detection constants below are sensible defaults that must be tuned on a
/// real device with real chanting.
final class VoiceJapaCounter: ObservableObject {

    enum Phase: Equatable { case idle, calibrating, listening, paused, denied }

    @Published private(set) var count = 0
    @Published private(set) var phase: Phase = .idle
    @Published private(set) var level: Float = 0        // 0...1, for the live meter
    @Published private(set) var calibrationProgress: Double = 0   // 0...1
    /// 0 (only loud, clear chanting) ... 1 (very sensitive, soft murmur). Default mid.
    @Published var sensitivity: Double = 0.55 { didSet { sensSnapshot = Float(sensitivity) } }

    /// Fired on the main thread each time a repetition is counted (for haptics/pulse).
    var onCount: ((Int) -> Void)?
    /// Fired once when `count` first reaches `target`.
    var onTargetReached: (() -> Void)?
    var target: Int = 108

    // MARK: - Audio-thread state (touched only inside the tap callback)
    private let engine = AVAudioEngine()
    private var noiseFloor: Float = 0.015
    private var isVoiced = false
    private var lastCountAt: CFTimeInterval = 0
    private var reachedTarget = false
    private var sensSnapshot: Float = 0.55        // main writes, audio reads (benign)

    // Calibration accumulators
    private var calibrating = false
    private var calibPeak: Float = 0
    private var calibStartAt: CFTimeInterval = 0
    private let calibDuration: CFTimeInterval = 6.0

    // Detection tuning (device-tuned later)
    private let refractory: CFTimeInterval = 0.34   // min seconds between counts
    private let floorAdapt: Float = 0.05            // how fast the noise floor tracks quiet

    var isRunning: Bool { phase == .listening || phase == .calibrating }

    // MARK: - Permission

    /// Requests mic permission and calls back on the main thread.
    func requestPermission(_ done: @escaping (Bool) -> Void) {
        AVAudioApplication.requestRecordPermission { granted in
            DispatchQueue.main.async {
                if !granted { self.phase = .denied }
                done(granted)
            }
        }
    }

    // MARK: - Session lifecycle

    func startListening() {
        requestPermission { ok in
            guard ok else { return }
            self.reachedTarget = false
            self.beginTap(calibrate: false)
            self.phase = .listening
        }
    }

    /// Runs the "chant 3 times" calibration, then calls `completion` on the main thread.
    func calibrate(_ completion: @escaping () -> Void) {
        requestPermission { ok in
            guard ok else { completion(); return }
            self.calibPeak = 0
            self.calibStartAt = CACurrentMediaTime()
            self.calibrationProgress = 0
            self.beginTap(calibrate: true)
            self.phase = .calibrating
            self.calibCompletion = completion
        }
    }
    private var calibCompletion: (() -> Void)?

    func pause() {
        guard phase == .listening else { return }
        engine.pause()
        phase = .paused
    }

    func resume() {
        guard phase == .paused else { return }
        try? engine.start()
        phase = .listening
    }

    func stop() {
        if engine.isRunning || !engine.inputNode.inputFormat(forBus: 0).sampleRate.isZero {
            engine.inputNode.removeTap(onBus: 0)
        }
        engine.stop()
        deactivateSession()
        if phase != .denied { phase = .idle }
    }

    func reset() {
        count = 0
        reachedTarget = false
    }

    func adjust(by delta: Int) {
        count = max(0, count + delta)
    }

    /// QA-only: preload a lifelike listening state for App Store screenshots.
    /// Only ever called behind the --voicejapa-demo launch argument (never in production);
    /// it starts no audio engine, so the sim needs no microphone.
    func loadDemoState(count demoCount: Int, target demoTarget: Int) {
        target = demoTarget
        count = demoCount
        level = 0.42
        phase = .listening
    }

    // MARK: - Engine

    private func beginTap(calibrate: Bool) {
        calibrating = calibrate
        configureSession()
        let input = engine.inputNode
        let format = input.inputFormat(forBus: 0)
        input.removeTap(onBus: 0)
        input.installTap(onBus: 0, bufferSize: 2048, format: format) { [weak self] buffer, _ in
            self?.process(buffer)
        }
        engine.prepare()
        do { try engine.start() }
        catch { DispatchQueue.main.async { self.phase = .denied } }
    }

    private func configureSession() {
        let s = AVAudioSession.sharedInstance()
        try? s.setCategory(.playAndRecord, mode: .measurement, options: [.duckOthers, .defaultToSpeaker])
        try? s.setActive(true, options: [])
    }

    private func deactivateSession() {
        try? AVAudioSession.sharedInstance().setActive(false, options: [.notifyOthersOnDeactivation])
    }

    // MARK: - Detection (audio thread)

    private func process(_ buffer: AVAudioPCMBuffer) {
        guard let ch = buffer.floatChannelData?[0] else { return }
        let n = vDSP_Length(buffer.frameLength)
        guard n > 0 else { return }

        var meanSquare: Float = 0
        vDSP_measqv(ch, 1, &meanSquare, n)
        let rms = sqrtf(meanSquare)

        let now = CACurrentMediaTime()

        if calibrating {
            calibPeak = max(calibPeak, rms)
            let p = min(1, (now - calibStartAt) / calibDuration)
            DispatchQueue.main.async {
                self.level = min(1, rms * 10)
                self.calibrationProgress = p
                if p >= 1 { self.finishCalibration() }
            }
            return
        }

        // Adaptive threshold from the learned noise floor + a sensitivity-scaled margin.
        // Higher sensitivity => smaller margin => softer chanting still counts.
        let margin = 0.006 + (1 - sensSnapshot) * 0.055
        let threshold = noiseFloor + margin

        // Track the noise floor only while quiet.
        if rms < threshold { noiseFloor = (1 - floorAdapt) * noiseFloor + floorAdapt * rms }

        var counted = false
        if !isVoiced && rms > threshold {
            isVoiced = true
            if now - lastCountAt > refractory {
                lastCountAt = now
                counted = true
            }
        } else if isVoiced && rms < threshold * 0.6 {   // hysteresis: clear well below
            isVoiced = false
        }

        DispatchQueue.main.async {
            self.level = min(1, rms * 10)
            if counted {
                self.count += 1
                self.onCount?(self.count)
                if !self.reachedTarget && self.count >= self.target {
                    self.reachedTarget = true
                    self.onTargetReached?()
                }
            }
        }
    }

    private func finishCalibration() {
        guard calibrating else { return }
        calibrating = false
        engine.inputNode.removeTap(onBus: 0)
        engine.stop()
        // Seed the noise floor a little under a fraction of the measured chant peak
        // so listening starts already tuned to this voice.
        if calibPeak > 0 { noiseFloor = max(0.008, calibPeak * 0.12) }
        phase = .idle
        let done = calibCompletion
        calibCompletion = nil
        deactivateSession()
        done?()
    }

    deinit {
        engine.inputNode.removeTap(onBus: 0)
        engine.stop()
    }
}

File: Android/BhaktiAngan/app/src/main/java/app/bhaktiangan/core/audio/VoiceJapaCounter.kt
package app.bhaktiangan.core.audio

import android.annotation.SuppressLint
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Handler
import android.os.Looper
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlin.concurrent.thread
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Hands-free japa counting, a 1:1 port of the iOS `VoiceJapaCounter` onset detector.
 *
 * Reads the mic with [AudioRecord], measures short-time loudness (RMS), and counts each
 * voiced burst (one mantra repetition separated by a breath). An adaptive noise floor +
 * hysteresis + a refractory window prevent double-counting and room-noise triggers. A
 * short calibration ("chant a few times") learns the devotee's level.
 *
 * Audio is processed live and is NEVER recorded or transmitted, so the Play Data-safety
 * "no data collected" holds. Foreground only. Detection constants mirror iOS and must be
 * tuned on a real device with real chanting.
 */
class VoiceJapaCounter {

    enum class Phase { IDLE, CALIBRATING, LISTENING, PAUSED, DENIED }

    private val _count = MutableStateFlow(0)
    val count: StateFlow<Int> = _count
    private val _phase = MutableStateFlow(Phase.IDLE)
    val phase: StateFlow<Phase> = _phase
    private val _level = MutableStateFlow(0f)
    val level: StateFlow<Float> = _level
    private val _calibProgress = MutableStateFlow(0f)
    val calibProgress: StateFlow<Float> = _calibProgress

    /** 0 (loud clear chanting only) ... 1 (soft murmur). Written on main, read on audio thread. */
    @Volatile var sensitivity: Float = 0.55f
    @Volatile var target: Int = 108

    var onCount: ((Int) -> Unit)? = null
    var onTargetReached: (() -> Unit)? = null

    // Detection state (audio thread only)
    private var noiseFloor = 0.015f
    private var isVoiced = false
    private var lastCountAtMs = 0L
    private var reachedTarget = false

    // Calibration
    private var calibrating = false
    private var calibPeak = 0f
    private var calibStartMs = 0L
    private var calibCompletion: (() -> Unit)? = null

    private val refractoryMs = 340L      // min gap between counts (iOS 0.34s)
    private val floorAdapt = 0.05f
    private val calibDurationMs = 6000L

    @Volatile private var running = false
    private var worker: Thread? = null
    private val main = Handler(Looper.getMainLooper())

    private val sampleRate = 16000
    private val minBuf = AudioRecord.getMinBufferSize(
        sampleRate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT,
    ).let { if (it <= 0) 4096 else it }

    fun startListening() {
        reachedTarget = false
        beginTap(calibrate = false, onCalibDone = null)
    }

    fun calibrate(onDone: () -> Unit) {
        calibPeak = 0f
        _calibProgress.value = 0f
        beginTap(calibrate = true, onCalibDone = onDone)
    }

    fun pause() { if (_phase.value == Phase.LISTENING) _phase.value = Phase.PAUSED }
    fun resume() { if (_phase.value == Phase.PAUSED) _phase.value = Phase.LISTENING }

    fun stop() {
        running = false
        worker?.join(400)
        worker = null
        if (_phase.value != Phase.DENIED) _phase.value = Phase.IDLE
    }

    fun reset() { _count.value = 0; reachedTarget = false }

    fun adjust(delta: Int) { _count.value = max(0, _count.value + delta) }

    /** QA-only: preload a lifelike listening state for Play screenshots (starts no audio). */
    fun loadDemoState(count: Int, target: Int) {
        this.target = target
        _count.value = count
        _level.value = 0.42f
        _phase.value = Phase.LISTENING
    }

    @SuppressLint("MissingPermission") // caller (screen) requests RECORD_AUDIO before starting
    private fun beginTap(calibrate: Boolean, onCalibDone: (() -> Unit)?) {
        if (running) stop()
        calibrating = calibrate
        calibCompletion = onCalibDone
        val record = try {
            AudioRecord(
                MediaRecorder.AudioSource.MIC, sampleRate,
                AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT, minBuf * 2,
            )
        } catch (t: Throwable) { _phase.value = Phase.DENIED; return }
        if (record.state != AudioRecord.STATE_INITIALIZED) { _phase.value = Phase.DENIED; runCatching { record.release() }; return }

        running = true
        _phase.value = if (calibrate) Phase.CALIBRATING else Phase.LISTENING
        calibStartMs = System.currentTimeMillis()
        val buf = ShortArray(2048)

        worker = thread(name = "voice-japa") {
            try {
                record.startRecording()
                while (running) {
                    if (_phase.value == Phase.PAUSED) { Thread.sleep(60); continue }
                    val n = record.read(buf, 0, buf.size)
                    if (n <= 0) continue
                    process(buf, n)
                }
            } catch (_: Throwable) {
                main.post { if (_phase.value != Phase.DENIED) _phase.value = Phase.DENIED }
            } finally {
                runCatching { record.stop() }
                runCatching { record.release() }
            }
        }
    }

    private fun process(buf: ShortArray, n: Int) {
        var sumSq = 0.0
        for (i in 0 until n) { val s = buf[i] / 32768f; sumSq += (s * s).toDouble() }
        val rms = sqrt(sumSq / n).toFloat()
        val now = System.currentTimeMillis()

        if (calibrating) {
            calibPeak = max(calibPeak, rms)
            val p = min(1f, (now - calibStartMs) / calibDurationMs.toFloat())
            _level.value = min(1f, rms * 10)
            _calibProgress.value = p
            if (p >= 1f) finishCalibration()
            return
        }

        // Adaptive threshold: noise floor + a sensitivity-scaled margin (higher sensitivity
        // => smaller margin => softer chanting still counts).
        val margin = 0.006f + (1 - sensitivity) * 0.055f
        val threshold = noiseFloor + margin
        if (rms < threshold) noiseFloor = (1 - floorAdapt) * noiseFloor + floorAdapt * rms

        var counted = false
        if (!isVoiced && rms > threshold) {
            isVoiced = true
            if (now - lastCountAtMs > refractoryMs) { lastCountAtMs = now; counted = true }
        } else if (isVoiced && rms < threshold * 0.6f) {
            isVoiced = false
        }

        _level.value = min(1f, rms * 10)
        if (counted) {
            val c = _count.value + 1
            _count.value = c
            main.post { onCount?.invoke(c) }
            if (!reachedTarget && c >= target) {
                reachedTarget = true
                main.post { onTargetReached?.invoke() }
            }
        }
    }

    private fun finishCalibration() {
        if (!calibrating) return
        calibrating = false
        running = false
        // Seed the noise floor under a fraction of the measured chant peak so listening
        // starts already tuned to this voice.
        if (calibPeak > 0) noiseFloor = max(0.008f, calibPeak * 0.12f)
        _phase.value = Phase.IDLE
        val done = calibCompletion
        calibCompletion = null
        main.post { done?.invoke() }
    }
}
